src/main/java/graphql/GraphQL.java

```diff
@@ -519,19 +521,26 @@ public ExecutionResult execute(ExecutionInput executionInput) {
         return execute(executionInputRef.get(), preparsedDoc.getDocument(), graphQLSchema, instrumentationState);
     }
 
-    private PreparsedDocumentEntry parseAndValidate(ExecutionInput executionInput, GraphQLSchema graphQLSchema, InstrumentationState instrumentationState) {
-        log.debug("Parsing query: '{}'...", executionInput.getQuery());
+    private PreparsedDocumentEntry parseAndValidate(AtomicReference<ExecutionInput> executionInputRef, GraphQLSchema graphQLSchema, InstrumentationState instrumentationState) {
+
+        ExecutionInput executionInput = executionInputRef.get();
+        String query = executionInput.getQuery();
+
+        log.debug("Parsing query: '{}'...", query);
         ParseResult parseResult = parse(executionInput, graphQLSchema, instrumentationState);
         if (parseResult.isFailure()) {
-            log.warn("Query failed to parse : '{}'", executionInput.getQuery());
+            log.warn("Query failed to parse : '{}'", query);
             return new PreparsedDocumentEntry(toInvalidSyntaxError(parseResult.getException()));
         } else {
             final Document document = parseResult.getDocument();
+            // they may have changed the document and the variables via instrumentation so update the reference to it
+            executionInput = executionInput.transform(builder -> builder.variables(parseResult.getVariables()));
```

Comment:
AFAIK any other instrumentation methods cannot change state. instrumentDocumentAndVariables feels inconsistent with the rest of Instrumentation interface. Should it not be a separate interface with Transformation in its name?

Reply:
Actually we have 6 - instrumentExecutionInput, DocumentAndVariables, Schema, ExecutionContext, DataFetcher, and ExecutionResult